graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/RenameFields.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2020 Graylog, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the Server Side Public License, version 1,
+ * as published by MongoDB, Inc.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.pipelineprocessor.functions.messages;
+
+import org.graylog.plugins.pipelineprocessor.EvaluationContext;
+import org.graylog.plugins.pipelineprocessor.ast.functions.AbstractFunction;
+import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionArgs;
+import org.graylog.plugins.pipelineprocessor.ast.functions.FunctionDescriptor;
+import org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor;
+import org.graylog.plugins.pipelineprocessor.rulebuilder.RuleBuilderFunctionGroup;
+import org.graylog2.plugin.Message;
+
+import java.util.Map;
+
+import static org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor.type;
+
+public class RenameFields extends AbstractFunction<Void> {
+
+    public static final String NAME = "rename_fields";
+
+    private final ParameterDescriptor<Map, Map> fieldsParam;
+    private final ParameterDescriptor<Message, Message> messageParam;
+
+    public RenameFields() {
+        fieldsParam = type("fields", Map.class).ruleBuilderVariable().description("The map of old name keys to new name values.").build();
+        messageParam = type("message", Message.class).optional().description("The message to use, defaults to '$message'").build();
+    }
+
+    @Override
+    public Void evaluate(FunctionArgs args, EvaluationContext context) {
+        final Map<String, String> fields = fieldsParam.required(args, context);
+        final Message message = messageParam.optional(args, context).orElse(context.currentMessage());
+
+        if (fields != null) {
+            for (String oldName : fields.keySet()) {
+                final String newName = fields.get(oldName);
+                if (message.hasField(oldName) && !oldName.equals(newName)) {
+                    message.addField(newName, message.getField(oldName));
+                    message.removeField(oldName);
+                }
```

Comment:
Some minor efficiency suggestions:
- Iterate over the entrySet of the Map so we can reduce the number of times we have to access it. This removes the repeated need for `fields.get()`
- Swap the check for old/new field name equality and `message.hasField()` check. This removes another map access of the message's fields if it isn't necessary.
```suggestion
            for (Map.Entry<String, String> entry : fields.entrySet()) {
                final String oldName = entry.getKey();
                final String newName = entry.getValue();
                
                if (!oldName.equals(newName) && message.hasField(oldName)) {
                    message.addField(newName, message.getField(oldName));
                    message.removeField(oldName);
                }
```

Reply:
Noice, just updated as suggested.